crates/accelerate/src/equivalence.rs

```diff
@@ -0,0 +1,786 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use itertools::Itertools;
+
+use pyo3::exceptions::PyTypeError;
+use qiskit_circuit::parameter_table::ParameterUuid;
+use rustworkx_core::petgraph::csr::IndexType;
+use rustworkx_core::petgraph::stable_graph::StableDiGraph;
+use rustworkx_core::petgraph::visit::IntoEdgeReferences;
+
+use smallvec::SmallVec;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+use std::{error::Error, fmt::Display};
+
+use exceptions::CircuitError;
+
+use ahash::RandomState;
+use indexmap::{IndexMap, IndexSet};
+use pyo3::prelude::*;
+use pyo3::types::{PyDict, PyList, PySet, PyString};
+
+use rustworkx_core::petgraph::{
+    graph::{EdgeIndex, NodeIndex},
+    visit::EdgeRef,
+};
+
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::circuit_instruction::OperationFromPython;
+use qiskit_circuit::imports::{ImportOnceCell, QUANTUM_CIRCUIT};
+use qiskit_circuit::operations::Param;
+use qiskit_circuit::operations::{Operation, OperationRef};
+use qiskit_circuit::packed_instruction::PackedOperation;
+
+mod exceptions {
+    use pyo3::import_exception_bound;
+    import_exception_bound! {qiskit.circuit.exceptions, CircuitError}
+}
+pub static PYDIGRAPH: ImportOnceCell = ImportOnceCell::new("rustworkx", "PyDiGraph");
+
+// Custom Structs
+
+#[pyclass(frozen, sequence, module = "qiskit._accelerate.equivalence")]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Key {
+    #[pyo3(get)]
+    pub name: String,
+    #[pyo3(get)]
+    pub num_qubits: u32,
```

Comment:
```suggestion
pub(crate) struct Key {
    #[pyo3(get)]
    pub(crate) name: String,
    #[pyo3(get)]
    pub(crate) num_qubits: u32,
```
I think visibility should be low by default.

Reply:
Key in this case should stay as a public struct because the function `EquivalenceLibrary::keys()` returns an iterator of `Key` instances, I can make the method public only to the crate as well but it is a little inconvenient since `EquivalenceLibrary::node_index(key)` is also a public method that in this case accepts `Key` as an argument. I could use `map` to turn them into tuple instances of `(&str, u32)` in both cases. However, this will require a `Clone` in the case of `EquivalenceLibrary::node_index(key)` since the `Key` struct stores a `String` instanced which would require for me to call `to_owned` on the `&str` instance.